modules/seqexec/server/src/main/scala/seqexec/server/ghost/GHOSTController.scala

```diff
@@ -22,10 +22,85 @@ import squants.time.Time
+  // 2. We observe in high resolution mode using only IFU1 on a target.
+  // Question: Do we need to, for (2), touch IFU2 at all (i.e. to standard mode, no target)?
+  // Do we need to always set target type for IFUs? How about bundle configuration?
+  //
+  // Right now, this only generate a config if the name, RA, and dec are defined: otherwise, we get the empty config.
+  private def ifuConfig(ifuNum: IFUNum,
+                        nameOpt: Option[String],
+                        raOpt: Option[HourAngle],
+                        decOpt: Option[Angle],
+                        bundleConfig: BundleConfig): Configuration =
+      (for {
+        name <- nameOpt
+        ra   <- raOpt
+        dec  <- decOpt
+      } yield {
+        Configuration.single(ifuNum.ifuStr + ".target", IFUTargetType.determineType(nameOpt).targetType) |+|
+          Configuration.single(ifuNum.ifuStr + ".ra", ra.toDoubleDegrees) |+|
+          Configuration.single(ifuNum.ifuStr + ".dec", dec.toDoubleDegrees) |+|
+          Configuration.single(ifuNum.ifuStr + ".bundle", bundleConfig.configName)
+      }).getOrElse(Configuration.Zero)
+
+  private def srifuConfig(config: GHOSTConfig): Configuration =
+    ifuConfig(IFU1, config.srifu1Name, config.srifu1CoordsRAHMS, config.srifu1CoordsDecDMS, StandardBundleConfig) |+|
+    ifuConfig(IFU2, config.srifu2Name, config.srifu2CoordsRAHMS, config.srifu2CoordsDecDMS, StandardBundleConfig)
+
+  private def hrifuConfig(config: GHOSTConfig): Configuration =
+    ifuConfig(IFU2, config.hrifu1Name, config.hrifu1CoordsRAHMS, config.hrifu1CoordsDecDMS, HighResBundleConfig) |+|
+      ifuConfig(IFU2, config.srifu2Name, config.srifu2CoordsRAHMS, config.srifu2CoordsDecDMS, StandardBundleConfig)
+
+  // If the srifu parameters are defined, use them; otherwise, use the hrifu parameters.
+  // TODO: What do we do with the base position explicit override?
+  // TODO: This was not on the list of provided parameter names.
+  private
+  def ghostConfig(config: GHOSTConfig): SeqActionF[F, CommandResult] = {
+    val giapiApply = srifuConfig(config) |+| hrifuConfig(config)
+
+    EitherT(ghostClient.genericApply(giapiApply).attempt)
+      .leftMap {
+        // The GMP sends these cryptic messages but we can do better
+        case CommandResultException(_, "Message cannot be null") => Execution("Unhandled Apply command")
+        case CommandResultException(_, m)                        => Execution(m)
+        case f                                                   => SeqexecException(f)
+      }
+  }
 
   def applyConfig(config: GHOSTConfig): SeqActionF[F, Unit] =
-    SeqActionF.apply(log.info(s"Configuring $config"))
+    for {
+      _ <- EitherT.liftF(Sync[F].delay(Log.debug("Start GHOST configuration")))
+      _ <- EitherT.liftF(Sync[F].delay(Log.debug(s"GHOST configuration $config")))
+      _ <- ghostConfig(config)
+      _ <- EitherT.liftF(
+        Sync[F].delay(Log.debug("Completed GHOST configuration")))
```

Comment:
I would factor out `EitherT.liftF(Sync[F].delay(Log.debug(message)))`. Is there not an existing lifted logger already?

Reply:
There is `SeqActionF.apply`

Reply:
I have to admit that I just snagged this right off of `GPIController` directly (with the name changes for GHOST, but that's it). I'm not entirely sure how this code works: I'm looking into @jluhrs' suggestion now and reading some Cats docs.

Reply:
Excellent. Thanks, @jluhrs. That cleans it up considerably and I understand a lot better how this code is working now. We should make the same change to `GPIController` to clean it up. I'll add a JIRA task and do that quickly.